File: day11.lua
#!/usr/bin/env lua

local memo = {}
for i = 0, 75 do memo[i] = {} end

local function blink(x, n)
    if memo[n][x] then return memo[n][x] end
    local y
    if n == 0 then y = 1
    elseif x == 0 then y = blink(1, n - 1)
    else
        local buf = tostring(x)
        local len = #buf
        if len % 2 == 0 then
            y = blink(tonumber(buf:sub(len // 2 + 1)), n - 1)
            y = y + blink(tonumber(buf:sub(1, len // 2)), n - 1)
        else
            y = blink(x * 2024, n - 1)
        end
    end
    memo[n][x] = y
    return y
end

local file = io.open("day11.txt")
local sum = { 0, 0 }

for t in file:read():gmatch("%d+") do
    local x = tonumber(t)
    sum[1] = sum[1] + blink(x, 25)
    sum[2] = sum[2] + blink(x, 75)
end

print(sum[1], sum[2])

File: day11.c
#include <stdio.h>
#include <stdlib.h>

#define M (1 << 20)
long long memo[M][75 + 1];

long long blink(long long x, int n)
{
    if (x < M && memo[x][n]) return memo[x][n];
    long long y;
    if (n == 0) y = 1;
    else if (x == 0) y = blink(1, n - 1);
    else {
        char buf[32];
        int len = sprintf(buf, "%lld", x);
        if (!(len % 2)) {
            y = blink(atoll(buf + len / 2), n - 1);
            buf[len / 2] = 0;
            y += blink(atoll(buf), n - 1);
        } else {
            y = blink(x * 2024, n - 1);
        }
    }
    if (x < M) memo[x][n] = y;
    return y;
}

int main()
{
    FILE* file = fopen("day11.txt", "r");
    long long sum = 0, sum2 = 0;

    for (int x; 1 == fscanf(file, "%d", &x); ) {
        sum += blink(x, 25);
        sum2 += blink(x, 75);
    }

    printf("%lld %lld", sum, sum2);
}
